ml-agents/mlagents/trainers/rl_trainer.py

```diff
@@ -72,6 +143,89 @@ def advance(self) -> None:
         """
         Steps the trainer, taking in trajectories and updates if ready
         """
-        super().advance()
+        with hierarchical_timer("process_trajectory"):
+            for traj_queue in self.trajectory_queues:
+                try:
+                    t = traj_queue.get_nowait()
+                    self._process_trajectory(t)
+                except AgentManagerQueue.Empty:
+                    pass
+        if self.training_progress < 1.0:
+            if self._is_ready_update():
+                with hierarchical_timer("update_policy"):
+                    self._update_policy()
+                    for q in self.policy_queues:
+                        # Get policies that correspond to the policy queue in question
+                        q.put(self.get_policy(q.behavior_id))
         if not self.is_training:
             self.clear_update_buffer()
+
+    @property
+    def training_progress(self) -> float:
+        """
+        Returns a float between 0 and 1 indicating how far along in the training progress the Trainer is.
+        If 1, the Trainer wasn't training to begin with, or max_steps
+        is reached.
+        """
+        if self.is_training:
+            return min(self.step / float(self.trainer_parameters["max_steps"]), 1.0)
+        else:
+            return 1.0
+
+    def write_tensorboard_text(self, key: str, input_dict: Dict[str, Any]) -> None:
+        """
+        Saves text to Tensorboard.
+        Note: Only works on tensorflow r1.2 or above.
+        :param key: The name of the text.
+        :param input_dict: A dictionary that will be displayed in a table on Tensorboard.
+        """
+        try:
+            with tf.Session() as sess:
+                s_op = tf.summary.text(
+                    key,
+                    tf.convert_to_tensor(
+                        ([[str(x), str(input_dict[x])] for x in input_dict])
+                    ),
+                )
+                s = sess.run(s_op)
+                self.stats_reporter.write_text(s, self.step)
+        except Exception:
+            LOGGER.info("Could not write text summary for Tensorboard.")
+            pass
+
+    def save_model(self, name_behavior_id: str) -> None:
```

Comment:
How come we are moving save_model/export_model out of trainer.py?

Reply:
This is probably a mistake - they should be abstract methods in trainer.py